XMonad/Hooks/DynamicLog.hs

```diff
@@ -585,6 +589,103 @@ pprWindowSetXinerama ws = "[" ++ unwords onscreen ++ "] " ++ unwords offscreen
+         }
+
+-- | Create an 'IconSet' from a 'String'
+appIcon :: String -> IconSet
+appIcon x = pure [baseIconSet x]
+
+-- | Dynamically changes 'WorkspaceId' based on the 'Window's inside the Workspace.
+-- 'IconSet's describe which icons are shown depending on which windows fit a query
+-- To create an 'IconSet' make a 'Query' that returns a '[Icon]'.
+-- 'appIcon' can be used to simplify this process
+-- For example,
+--
+-- > icons :: IconSet
+-- > icons = composeAll
+-- >           [ className =? "discord" --> appIcon "\xfb6e"
+-- >           , className =? "Discord" --> appIcon "\xf268"
+-- >           , className =? "Firefox" --> appIcon "\63288"
+-- >           , className =? "Spotify" <||>  className =? "spotify" --> appIcon "阮"
+-- >           ]
+--
+-- then you can add the hook to your config
+--
+-- > xmonad $ def
+-- >      {
+-- >      logHook = dynamicLogIconsWithPP icons xmobarPP <> myManageHook
+-- >      }
+--
+dynamicLogIconsWithPP :: IconSet -- ^ The 'IconSet' to use
+                      -> PP -- ^ The 'PP' to alter
+                      -> X () -- ^ The resulting 'X' action
+dynamicLogIconsWithPP iconset pp = dynamicLogWithPP =<< dynamicLogIconsConvert (def{ iconConfigIcons = iconset, iconConfigPP = pp})
+
+-- | Datatype for expanded 'Icon' configurations
+data IconConfig = IconConfig
+    { iconConfigIcons :: IconSet -- ^ The 'IconSet' to use
+    , iconConfigStack :: [String] -> String -- ^ The function to manage stacking of 'Icon's
+    , iconConfigPP :: PP -- ^ The 'PP' to alter
+    }
+
+instance Default IconConfig where
+    def = IconConfig
+        { iconConfigIcons = mempty
+        , iconConfigStack = wrap "[" "]" . unwords
+        , iconConfigPP = def
+        }
+
+-- | This is the same as 'dynamicLogIconsWithPP' but it takes a 'IconConfig'.
+-- This allows you to manually customise the 'Icon's the stacking function and also your `PP`
+dynamicLogIconsConvert :: IconConfig -> X PP
+dynamicLogIconsConvert iconConfig = do
+    ws <- gets (S.workspaces . windowset)
+    icons <- M.fromList . (maybeToList =<<) <$> mapM (getIcons (iconConfigIcons iconConfig)) ws
+    pure $ (iconConfigPP iconConfig)
+        { ppCurrent =  ppSection ppCurrent iconCurrent icons
+        , ppVisible =  ppSection ppVisible iconVisible icons
+        , ppHidden = ppSection ppHidden iconHidden icons
+        , ppHiddenNoWindows =  ppSection ppHiddenNoWindows iconHiddenNoWindows icons
+        }
+  where
+    ppSection pF f icons = pF (iconConfigPP iconConfig) . concatIcons f . iconLookup icons
+    iconLookup icons x = M.findWithDefault [baseIconSet x] x icons
+    concatIcons f y
+        | length y > 1 = iconConfigStack iconConfig $ map f y
+        | otherwise = concatMap f y
+
+getIcons :: IconSet -> WindowSpace -> X (Maybe (WorkspaceId,[Icon]))
+getIcons is w = do
+    a <- sequence $ foldMap (runQuery is) . S.integrate <$> S.stack w
+    pure $ (S.tag w,) <$> (a >>= \x -> if null x then Nothing else Just  x)
```

Comment:
more minor spacing fixes. Additionally, instead of the `a >>= \x -> ...` you could use Control.Monad.mfilter, making it a tiny bit more readable (as that more clearly highlights the semantic meaning of that bit of code)
```suggestion
getIcons :: IconSet -> WindowSpace -> X (Maybe (WorkspaceId, [Icon]))
getIcons is w = do
    a <- sequence $ foldMap (runQuery is) . S.integrate <$> S.stack w
    pure $ (S.tag w,) <$> (a >>= \x -> if null x then Nothing else Just x)
```

Reply:
also, `a` is quite meaningless of a name here 😁